File: LeetCode/1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        int open = 0;
        string temp = "";
        
        for(char c:s) {
            if(c == '(') {
                open++;
            } else if(c == ')') {
                if(open == 0)
                    continue;
                open--;
            }
            
            temp.push_back(c);
        }
        
        int n = temp.length();
        string result = "";
        for(int i = n-1; i>=0; i--) {
            if(temp[i] == '(' && open-- > 0)
                continue;
            result.push_back(temp[i]);
        }
        
        reverse(result.begin(), result.end());
        return result;
    }
};

// class Solution {
// public:
//     string minRemoveToMakeValid(string s) {
//         int open = 0;
//         int n = s.length();
//         string ans = "";
//         for (int i = 0; i < n; i++) {
//             if (s[i] >= 'a' && s[i] <= 'z') {
//                 ans.push_back(s[i]);
//             } else if (s[i] == '(') {
//                 open++;
//                 ans.push_back(s[i]);
//             } else if (open > 0) {
//                 open--;
//                 ans.push_back(s[i]);
//             }
//         }
//         string result = "";
//         int close = 0;
//         n = ans.length();
//         for (int i = n - 1; i >= 0; i--) {
//             if (ans[i] >= 'a' && ans[i] <= 'z') {
//                 result.push_back(ans[i]);
//             } else if (ans[i] == ')') {
//                 close++;
//                 result.push_back(ans[i]);
//             } else if (close > 0) {
//                 close--;
//                 result.push_back(ans[i]);
//             }
//         }
//         reverse(begin(result), end(result));
//         return result;
//     }
// };

// // class Solution {
// // public:
// //     string minRemoveToMakeValid(string s) {
// //         stack<int> st;
// //         int n = s.length();
// //         unordered_set<int> remove_idx;
// //         for (int i = 0; i < n; i++) {
// //             if (s[i] == '(') {
// //                 st.push(i);
// //             } else if (s[i] == ')') {
// //                 if (st.empty()) {
// //                     remove_idx.insert(i);
// //                 } else {
// //                     st.pop();
// //                 }
// //             }
// //         }
// //         while (!st.empty()) {
// //             remove_idx.insert(st.top());
// //             st.pop();
// //         }
// //         string ans = "";
// //         for (int i = 0; i < n; i++) {
// //             if (remove_idx.find(i) == remove_idx.end()) {
// //                 ans.push_back(s[i]);
// //             }
// //         }
// //         return ans;
// //     }
// // };

File: LeetCode/1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        result = []
        n = len(s)

        last_open = 0
        for i in range(n):
            c = s[i]
            if c.islower():
                result.append(c)
            elif c == '(':
                result.append(c)
                last_open += 1
            elif last_open > 0:
                last_open -= 1
                result.append(c)

        if not result:
            return ""

        s = ''.join(result)
        result = []
        last_close = 0
        n = len(s)

        for i in range(n - 1, -1, -1):
            c = s[i]
            if c.islower():
                result.append(c)
            elif c == ')':
                result.append(c)
                last_close += 1
            elif last_close > 0:
                last_close -= 1
                result.append(c)

        return ''.join(result[::-1])


# class Solution:
#     def minRemoveToMakeValid(self, s: str) -> str:
#         open_count = 0
#         temp = ""

#         for c in s:
#             if c == '(':
#                 open_count += 1
#             elif c == ')':
#                 if open_count == 0:
#                     continue
#                 open_count -= 1

#             temp += c

#         n = len(temp)
#         result = ""
#         for i in range(n - 1, -1, -1):
#             if temp[i] == '(' and open_count > 0:
#                 open_count -= 1
#                 continue
#             result += temp[i]

#         return result[::-1]
        
# # class Solution:
# #     def minRemoveToMakeValid(self, s: str) -> str:
# #         st = []
# #         n = len(s)
# #         idx = set()
# #         for i in range(n):
# #             if s[i] == '(':
# #                 st.append(i)
# #             elif s[i] == ')':
# #                 if st and s[st[-1]] == '(':
# #                     st.pop()
# #                 else:
# #                     idx.add(i)
# #         idx.update(st)
# #         ans = ""
# #         for i in range(n):
# #             if i not in idx:
# #                 ans += s[i]
# #         return ans